README.md

[![Travis-CI Build Status](https://travis-ci.org/mdsumner/spdplyr.svg?branch=master)](https://travis-ci.org/mdsumner/spdplyr) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/mdsumner/spdplyr?branch=master&svg=true)](https://ci.appveyor.com/project/mdsumner/spdplyr) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/spdplyr)](https://cran.r-project.org/package=spdplyr) [![Coverage Status](https://img.shields.io/codecov/c/github/mdsumner/spdplyr/master.svg)](https://codecov.io/github/mdsumner/spdplyr?branch=master)

<!-- README.md is generated from README.Rmd. Please edit that file -->
spdplyr
=======

`spdplyr` provides methods for the `dplyr` verbs for the `Spatial` classes in the `sp` package.

Installation
------------

Install from CRAN:

``` r
install.packages("spdplyr")
```

The development version of `spdplyr` can be installed directly from Github:

``` r
devtools::install_github("mdsumner/spdplyr")
```

dplyr verbs for Spatial
=======================

Apply `dplyr` verbs to the attribute data of `sp` objects with dplyr verbs.

See `?dplyr-Spatial'` for supported verbs.

How does it work?
-----------------

Most of it is easy, simply applying the analogous operation of the verb to the Spatial\*DataFrame and subsetting/re-organizing as necessary. Some operations use the `spbabel::sptable` approach which provides a loss-less way to do what `ggplot2::fortify` does and its reverse.

Issues
======

**NOTE:** joins aren't really thought through or developed properly, send me your feedback / ideas / help!


NEWS.md
# spdplyr 0.1.2

* updated to use sp rather than spFromTable in line with spbabel

* modified the print/show details for Spatial 

* removed some leftover references from original source

# spdplyr 0.1.1

* first working version 

* forked from spbabel 0b95adb3, previous news is from spbabel originally

* various improvements provided by jlegewie, removed transmute_ (not needed), improved filter_ and select_, added left_join and inner_join, see https://github.com/mdsumner/spbabel/pull/10

* added group_by and complementary summarize capability for Spatial 

* set data.frame and tbl and tbl_df as S4 compatible



tests/testthat.R
library(testthat)
library(spdplyr)

test_check("spdplyr")


tests/testthat/test-adv-dplyr.R
library(testthat)
library(dplyr)
context("adv-dplyr")


library(maptools)
library(dplyr)
data(wrld_simpl)
poly1 <- wrld_simpl
line1 <- as(wrld_simpl, "SpatialLinesDataFrame")
point1 <- as(line1, "SpatialPointsDataFrame")
test_that("group by and summarize is quiet", {
  testthat::skip_on_cran()
expect_silent(g <- wrld_simpl  %>% group_by(REGION)  %>% summarize(alon = mean(LON), mxlat = max(LAT), mxarea = max(AREA)))
})


#mpoint1 <- SpatialMultiPointsDataFrame(lapply(split(line1, seq(nrow(line1))), function(y) coordinates(as(y, "SpatialPoints"))), 
#                                       as.data.frame(line1))
# filter() (and slice())
# arrange()
# select() (and rename())
# distinct()
# mutate() (and transmute())
# summarise()
# sample_n() and sample_frac()
# 
# group_by and summarize
# 
# inner_join()	SELECT * FROM x JOIN y ON x.a = y.a
# left_join()	SELECT * FROM x LEFT JOIN y ON x.a = y.a
# right_join()	SELECT * FROM x RIGHT JOIN y ON x.a = y.a
# full_join()	SELECT * FROM x FULL JOIN y ON x.a = y.a
# semi_join()	SELECT * FROM x WHERE EXISTS (SELECT 1 FROM y WHERE x.a = y.a)
# anti_join()	SELECT * FROM x WHERE NOT EXISTS (SELECT 1 FROM y WHERE x.a = y.a)
# intersect(x, y)	SELECT * FROM x INTERSECT SELECT * FROM y
# union(x, y)	SELECT * FROM x UNION SELECT * FROM y
# setdiff(x, y)	SELECT * FROM x EXCEPT SELECT * FROM y
# 

test_that("everthing is ok", {
  testthat::skip_on_cran()
  x1 <- poly1
  
  x1 %>% filter(NAME == "Australia")
  x1 %>% filter(POP2005 > 2e6)
  x1 %>% filter(grepl("Aust", NAME))
  x1 %>% slice(c(100, 200))
  
  x1 %>% arrange(AREA)
  x1 %>% dplyr::select(NAME, POP2005, AREA)
  x1 %>% rename(Country = NAME, Population = POP2005)
  x1 %>% distinct(AREA, .keep_all = TRUE)
  x1 %>% mutate(AREA = REGION * 2)
  x1 %>% transmute(NAME = gsub("^A", "Z", NAME))
  x1 %>% summarize(a = 'POP2005')
  
  #x1 %>% group_by(REGION)
  #x1 %>% summarize(a = POP2005)
})

tests/testthat/test-basic-dplyr.R
library(testthat)
context("dplyr-Spatial")

## TODO: Rename context
## TODO: Add more tests

library(maptools)
library(spdplyr)
data(wrld_simpl)
poly1 <- wrld_simpl
line1 <- as(wrld_simpl, "SpatialLinesDataFrame")
point1 <- as(line1, "SpatialPointsDataFrame")
#mpoint1 <- SpatialMultiPointsDataFrame(lapply(split(line1, seq(nrow(line1))), function(y) coordinates(as(y, "SpatialPoints"))), 
#                                       as.data.frame(line1))

data(mpoint1)


test_that("filter works for all geometric types", {
  expect_that(nrow(filter(poly1, NAME == "Australia")), equals(1L) )
  expect_that(nrow(poly1 %>% filter(NAME == "Australia")), equals(1L) )
  
  expect_that(nrow(filter(line1, NAME == "Australia")), equals(1L) )
  expect_that(nrow(line1 %>% filter(NAME == "Australia")), equals(1L) )
  
  
  expect_that(nrow(filter(point1, NAME == "Australia")), equals(611L) )
  expect_that(nrow(point1 %>% filter(NAME == "Australia")), equals(611L) )
  
  expect_that(nrow(filter(mpoint1, NAME == "Australia")), equals(1L) )
  expect_that(nrow(mpoint1 %>% filter(NAME == "Australia")), equals(1L) )
  
  ## multipoint needs work, slicing here duplicates row.names so catch this warning as expected for now
  expect_that(nrow(slice(mpoint1, 1:2)), testthat::gives_warning("some row.names duplicated"))
})

# filter() (and slice())
# arrange()
# select() (and rename())
# distinct()
# mutate() (and transmute())
# summarise()
# sample_n() and sample_frac()

test_that("arrange works", {
  expect_that(nrow(arrange(poly1)), equals(246L) )
  expect_that(nrow(arrange(point1)), equals(26264L) )
  expect_that(arrange(line1)$UN, equals(c(28L, 12L, 31L, 8L, 51L, 24L, 16L, 32L, 36L, 48L, 52L, 60L, 
                                          44L, 50L, 84L, 70L, 68L, 104L, 204L, 90L, 76L, 100L, 96L, 124L, 
                                          116L, 144L, 178L, 180L, 108L, 156L, 4L, 64L, 152L, 136L, 120L, 
                                          148L, 174L, 170L, 188L, 140L, 192L, 132L, 184L, 196L, 208L, 262L, 
                                          212L, 214L, 218L, 818L, 372L, 226L, 233L, 232L, 222L, 231L, 40L, 
                                          203L, 254L, 246L, 242L, 238L, 583L, 258L, 250L, 270L, 266L, 268L, 
                                          288L, 308L, 304L, 276L, 316L, 300L, 320L, 324L, 328L, 332L, 340L, 
                                          191L, 348L, 352L, 356L, 364L, 376L, 380L, 384L, 368L, 392L, 388L, 
                                          400L, 404L, 417L, 408L, 296L, 410L, 414L, 398L, 418L, 422L, 428L, 
                                          112L, 440L, 430L, 703L, 438L, 434L, 450L, 474L, 496L, 500L, 807L, 
                                          466L, 504L, 480L, 478L, 470L, 512L, 462L, 484L, 458L, 508L, 454L, 
                                          540L, 570L, 562L, 533L, 660L, 56L, 344L, 580L, 234L, 20L, 292L, 
                                          833L, 442L, 446L, 492L, 275L, 499L, 175L, 248L, 574L, 166L, 10L, 
                                          74L, 260L, 334L, 86L, 162L, 581L, 548L, 566L, 528L, 578L, 524L, 
                                          520L, 740L, 558L, 554L, 600L, 604L, 586L, 616L, 591L, 620L, 598L, 
                                          624L, 634L, 638L, 642L, 498L, 608L, 630L, 643L, 646L, 682L, 659L, 
                                          690L, 710L, 426L, 72L, 686L, 705L, 694L, 702L, 706L, 724L, 662L, 
                                          736L, 752L, 760L, 756L, 780L, 764L, 762L, 772L, 776L, 768L, 678L, 
                                          788L, 792L, 798L, 795L, 834L, 800L, 826L, 804L, 840L, 854L, 858L, 
                                          860L, 670L, 862L, 92L, 704L, 850L, 516L, 876L, 882L, 748L, 887L, 
                                          894L, 716L, 360L, 312L, 530L, 784L, 626L, 612L, 585L, 584L, 666L, 
                                          654L, 674L, 796L, 732L, 688L, 336L, 744L, 663L, 652L, 831L, 832L, 
                                          239L, 158L)))
  
  expect_that(arrange(poly1, NAME, UN)$SUBREGION[100], equals(30L))
})



test_that("select works", {
  expect_that(names(select_(poly1, "ISO2", "NAME")), equals(c("ISO2", "NAME")))
  expect_that(names(select_(line1, "ISO2", "NAME")), equals(c("ISO2", "NAME")))

})

# filter() (and slice())
# arrange()
# select() (and rename())
# distinct()
# mutate() (and transmute())
# summarise()
# sample_n() and sample_frac()

test_that("distinct works", {
          expect_that(nrow(distinct(poly1, REGION, .keep_all = TRUE)), equals(6L))
  expect_that(distinct(poly1, REGION, .keep_all = TRUE), is_a("SpatialPolygonsDataFrame"))
  
          })

test_that("mutate works", {
  expect_silent(mutate(poly1, NAME = "all same", AREA = row_number()))
  expect_silent(transmute(line1, Country= NAME, LON = 100))
  expect_that(transmute(line1, Country= NAME, LON = 100), is_a("SpatialLinesDataFrame"))
})


test_that("various examples", {
  testthat::skip_on_cran()
  expect_silent(wrld_simpl %>% mutate(NAME = "allthesame", REGION = row_number()))
  expect_silent(wrld_simpl %>% transmute(alpha = paste0(FIPS, NAME)) )
  expect_silent(wrld_simpl %>% filter(NAME %in% c("New Zealand", "Australia", "Fiji")))
  expect_silent(wrld_simpl %>% arrange(LON) )
  expect_silent(wrld_simpl %>% slice(c(9, 100)))
  expect_silent(wrld_simpl %>% dplyr::select(UN, FIPS))
  expect_silent(wrld_simpl %>% rename(`TM_WORLD_BORDERS_SIMPL0.2NAME` = NAME))
  expect_silent(wrld_simpl %>% distinct(REGION, .keep_all = TRUE) %>% arrange(REGION) )  ## first alphabetically in REGION
  expect_silent(wrld_simpl %>% arrange(REGION, desc(NAME)) %>% distinct(REGION, .keep_all = TRUE)) ## last
   
  ## we don't need to use piping
  expect_silent( slice(filter(mutate(wrld_simpl, likepiping = FALSE), abs(LON - 5) < 35 & LAT > 50), 4))
 ## summarise/ze is different, we have to return only one geometry
  expect_that(wrld_simpl %>% summarize(max(AREA)), is_a(class(wrld_simpl)))
  expect_silent(as(wrld_simpl, "SpatialLinesDataFrame") %>% mutate(perim = rnorm(246)))
  }
  
 )



ws <- filter(wrld_simpl, NAME %in% c("Portugal", "France", "Spain"))
ws$chNAME <- levels(ws$NAME)[ws$NAME]
d <- tibble(ent = c("Australia", "Portugal", "New Zealand", "France", "Spain"), AREA = seq(1.2, 2, length = 5))
test_that("joins work", {
  expect_that(all(is.na(left_join(ws@data, d)$ent)), is_true())
  expect_that(all(is.na(left_join(ws, d)$ent)), is_true())
  
  expect_that(left_join(ws@data, d, c("NAME" = "ent")), gives_warning("joining character"))
  expect_that(left_join(ws, d, c("chNAME" = "ent")), is_a(class(ws)))
  
  expect_that(nrow(inner_join(ws@data, d)), equals(0L))
  ## we can get 0-row S*DF
  expect_that(nrow(inner_join(ws, d)), equals(0L))
  
  expect_that(inner_join(ws@data, d, c("NAME" = "ent")), gives_warning("joining factor and character"))
  expect_that(inner_join(ws, d, c("chNAME" = "ent")), is_a(class(ws)))
  
  expect_that(nrow(inner_join(ws, d, c("chNAME" = "ent"))), equals(3L))
  
  expect_that(nrow(inner_join(ws, d[1:4, ], c("chNAME" = "ent"))), equals(2L))
  
})


R/utils.r

has_data <- function(x) .hasSlot(x, "data")


#' @importFrom sp bbox proj4string 
#' @importFrom dplyr  group_by
#' @importFrom tibble as_tibble
.print_Spatial <- 
function (x, ...) 
{
  cat("class       :", class(x), "\n")
  hasData <- FALSE
  nc <- 0

  if (has_data(x)) {
    hasData <- TRUE
    nc <- ncol(x@data)
  }
  ln <- 1
  ln <- length(x)
  cat("features    :", ln, "\n")

  e <- bbox(x)
  if (ln > 0) {
    cat("extent      : ", e[1, 1], ", ", e[1, 2], ", ", e[2, 
                                                          1], ", ", e[2, 2], "  (xmin, xmax, ymin, ymax)\n", 
        sep = "")
  }
  cat("coord. ref. :", proj4string(x), "\n")
  if (hasData) {
    x <- x@data
    cat("variables   : ", nc, "\n", sep = "")
    if (!inherits(x, "tbl_df")) x <- as_tibble(x)
    print(x)
  } 
  invisible(NULL)
}

#' Sp methods
#' @param object Spatial object
#' @title sp methods
#' @importFrom methods show .hasSlot
#' @rdname sp-methods
#' @export
setMethod ('show' , 'Spatial', 
           function(object) {
             .print_Spatial(object)
           }
)
#' @rdname sp-methods
#' @export
setMethod ('show' , 'SpatialPoints', 
           function(object) {
             .print_Spatial(object)
           }
)


release_questions <- 
  c("Have you single-quoted package and other non-English names in the DESCRIPTION and TITLE?\nhttps://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file", 
  c("Have you manually inspected your rendered vignettes?"))
    
    

R/sdplyr-package.r
#' Data Manipulation Verbs for Spatial Classes.
#'
#' @aliases NULL
#' @details The spdplyr package provides methods for the dplyr verbs. 
#' @section O. dplyr verbs:
#'  \tabular{ll}{
#'   \code{\link[dplyr]{filter}} \tab  filter   \cr
#'   \code{\link[dplyr]{slice}} \tab   slice   \cr
#'   \code{\link[dplyr]{arrange}} \tab arrange   \cr
#'   \code{\link[dplyr]{select}} \tab  select   \cr
#'   \code{\link[dplyr]{rename}} \tab  rename   \cr
#'   \code{\link[dplyr]{distinct}} \tab distinct  \cr
#'   \code{\link[dplyr]{mutate}} \tab   mutate  \cr
#'   \code{\link[dplyr]{transmute}} \tab transmute \cr
#'   \code{\link[dplyr]{group_by}} \tab group_by \cr
#'   \code{\link[dplyr]{summarise}} \tab summarise (or summarize) \cr
#'  }
#'  
#' @section I. dplyr-Spatial: dplyr verbs operating directly on Spatial objects
#' The implementation of \code{group_by} requires that sp objects are able use the object extending class "data.frame", 
#' \code{summarise} acts on the attributes in the table according to the expressions use and on the geometry by performing a non-topological union. 
#' @name spdplyr-package
#' @docType package
NULL


#' MultiPointsDataFrame data set
#' 

#' @name mpoint1
#' @docType data
#' @rdname mpoint1
NULL

#' "South-east" map data. 
#' 
#' Created in /data-raw/ \code{spmap} is a subset of wrld_simpl from maptools. 
#' @name spmap
#' @docType data
#' @examples 
#' library(dplyr)
#' spmap %>% filter(NAME == "Antarctica")
NULL



R/sdplyr-Spatial.r
setOldClass( c("tbl_df", "tbl", "data.frame" ) )
setOldClass( c("grouped_df", "tbl_df", "tbl", "data.frame" ) )

#' Dplyr verbs for Spatial
#' 
#' Direct application of the dplyr verbs to Spatial objects. There is no need for a conversion from and to Spatial with this approach. Not all verbs are supported, see Details. 
#' 
#' mutate, transmute, filter, arrange, slice, select, rename, distinct all work with attributes on the "data" slot and leave the geometry unchanged. 
#' 
#' summarise collapses to a grouped geometries by listing all subgeometries together, it does not perform any topological union or merge, and it takes no account of the calculations done on attributes. 
#' This is a brutal collapse of all the data, and is identical to what is seen with spplot(x, "group"). The behaviour of geometric collapse like this
#' is touch and go anyway, see the examples for a what `rgeos::gUnion` does. 
#' 
#'  summarise for points and multipoints, ... todo single Multipoint for multiple points
#' @param .data A tbl.
#' @param ... Name-value pairs of expressions. See \code{\link[dplyr]{mutate_}}
#' @param .dots Used to work around non-standard evaluation. 
#' @note Beware that attributes stored on Spatial objects *are not* linked to the geometry. Attributes are often used to store the area or perimeter length or centroid values but these may be completely unmatched to the underlying geometries. 
#' @rdname dplyr-Spatial
#' @name dplyr-Spatial
#' @export
#' @examples 
#' library(sp)
#' library(maptools)
#' data(wrld_simpl)
#' library(spdplyr) 
#' library(raster)  
#' wrld_simpl %>% mutate(NAME = "allthesame", REGION = row_number())
#' wrld_simpl %>% transmute(alpha = paste0(FIPS, NAME))
#' wrld_simpl %>% filter(NAME %in% c("New Zealand", "Australia", "Fiji"))
#' \dontrun{
#' wrld_simpl %>% arrange(LON)
#' wrld_simpl %>% slice(c(9, 100))
#' wrld_simpl %>% dplyr::select(UN, FIPS)
#' wrld_simpl %>% rename(`TM_WORLD_BORDERS_SIMPL0.2NAME` = NAME)
#' wrld_simpl %>% distinct(REGION, .keep_all = TRUE) %>% 
#'    arrange(REGION)  ## first alphabetically in REGION
#' wrld_simpl %>% arrange(REGION, desc(NAME)) %>% distinct(REGION, .keep_all = TRUE) ## last
#' }
#' ## we don't need to use piping
#' slice(filter(mutate(wrld_simpl, likepiping = FALSE), abs(LON - 5) < 35 & LAT > 50), 4)
#' 
#' 
#' ## works with Lines
#' #as(wrld_simpl, "SpatialLinesDataFrame") %>% 
#'  # mutate(perim = rgeos::gLength(wrld_simpl, byid = TRUE))
#' 
#' \dontrun{
#' ## summarise/ze can be used after group_by, or without
#' wrld_simpl %>% filter(REGION == 150) %>% summarize(max(AREA)) 
#' wrld_simpl %>% group_by(REGION) %>% summarize(max(AREA)) %>% 
#' plot(col = rainbow(nlevels(factor(wrld_simpl$REGION)), alpha = 0.3))
#' }
#' @importFrom dplyr %>% arrange mutate_ transmute_ filter_ arrange_ slice_ select_ rename_ distinct_ summarise_
#' @importFrom lazyeval all_dots
mutate_.Spatial <-  function(.data, ..., .dots) {
  dots <- lazyeval::all_dots(.dots, ..., all_named = TRUE)
  
  if (.hasSlot(.data, "data")) {
    dat <- mutate_(as.data.frame(.data), .dots = dots)
  } else {
    stop("no data to mutate for a %s", class(.data))
  }
  .data@data <- dat
  .data
}


#' @rdname dplyr-Spatial
#' @export
#' @importFrom dplyr inner_join
#' @importFrom spbabel sptable sp 
#' @importFrom tibble tibble
#' @examples 
#' ## group_by and summarize
#' 
#' \dontrun{
#' g <- wrld_simpl  %>% group_by(REGION)  %>% 
#'  summarize(alon = mean(LON), mxlat = max(LAT), mxarea = max(AREA))
#' g %>% mutate(ar = factor(REGION)) %>% spplot("ar")
#' w <- wrld_simpl
#' w$ar <- factor(w$REGION)
#' spplot(w, "ar")
#' }
#' \dontrun{
#' # compare what rgeos gives
#' ##spplot(rgeos::gUnionCascaded(w, id = w$ar))  ## good grief, is this compelling . . .
#' ## this is hardly a clean dissolve
#' ##plot(rgeos::gUnionCascaded(w, id = w$ar), col = rainbow(nlevels(factor(w$ar)), alpha = 0.5))
#' }
summarise_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data for distinct for a %s", class(.data))
  }
  # this should only ever return one-row objects
  # we cannot group_by on Spatial
  dat <- summarise_(as.data.frame(.data), ...)
  
  # row.names(dat) <- "1"
  gbomb <- spbabel::sptable(.data)
  if (inherits(.data@data, "grouped_df")) {
    groups <- attr(.data@data, "indices")  ## only robust for single-level group_by for now
    regroup <- tibble(labs =  unlist(lapply(seq_along(attr(.data@data, "group_sizes")), function(x) rep(x, attr(.data@data, "group_sizes")[x]))), 
                          inds = unlist(groups) + 1)
    
    gbomb <- gbomb  %>% 
      inner_join(regroup, c("object_" = "inds"))  %>% 
      mutate_(object_ = quote(labs))  %>% 
      select_(quote(-labs))  %>% 
      arrange_("object_", "branch_", "order_")
  
    } else {
    gbomb$object_ <- 1
  }
  spbabel::sp(gbomb, attr_tab = dat, crs = proj4string(.data))
  
}

#' @importFrom dplyr group_by_
#' @importFrom tibble as_tibble 
#' @rdname dplyr-Spatial
#' @export
group_by_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data for distinct for a %s", class(.data))
  }
  orownames <- row.names(.data)
  dat <- group_by_(as_tibble(as.data.frame(.data)), ...)
  
  #groupatts <- attributes(dat)
  #groupatts$class <- "data.frame"
  #groupatts$row.names <- orownames
  #dat <- as.data.frame(dat)
  #attributes(dat) <- groupatts
  .data@data <- dat
  .data
}








#' @rdname dplyr-Spatial
#' @export
filter_.Spatial <- function(.data, ..., .dots) {
  if (!.hasSlot(.data, "data")) {
    stop("no data to filter for a %s", class(.data))
  }
  dots <- lazyeval::all_dots(.dots, ..., all_named = TRUE)
  masks <- lazyeval::lazy_eval(dots, data = as.data.frame(.data@data))
  subset(.data, Reduce(`&`, masks))
}


#' @rdname dplyr-Spatial
#' @importFrom tibble as_tibble 
#' @importFrom dplyr arrange_ distinct_ rename_ select_ slice_ filter_ transmute_ mutate_ 
#' @export
arrange_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data to arrange for a %s", class(.data))
  }
  dat <- as_tibble(as.data.frame(.data))
  dat$order <- seq(nrow(dat))
  dat <- arrange_(dat, ...)
  .data[dat$order, ]
}


#' @rdname dplyr-Spatial
#' @importFrom tibble as_tibble 
#' @export
slice_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data to slice for a %s", class(.data))
  }
  dat <-  as_tibble(as.data.frame(.data))
  dat$order <- seq(nrow(dat))
  dat <- slice_(dat, ...)
  .data[dat$order, ]
}

#' @rdname dplyr-Spatial
#' @export
select_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data to select for a %s", class(.data))
  }
  dat <-  select_(as.data.frame(.data), ...)
  .data[, names(dat)]
}

#' @rdname dplyr-Spatial
#' @importFrom tibble as_tibble 
#' @export
rename_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data to rename for a %s", class(.data))
  }
  onames <- names(.data)
  dat <-  rename_(as_tibble(as.data.frame(.data)), ...)
  names(.data) <- names(dat)
  .data
}


#' @importFrom utils tail
#' @importFrom tibble as_tibble
#' @rdname dplyr-Spatial
#' @export
distinct_.Spatial <- function(.data, ...) {
  if (!.hasSlot(.data, "data")) {
    stop("no data for distinct for a %s", class(.data))
  }
  #orownames <- rownames(.data)
  nam <- utils::tail(make.names(c(names(.data), "order"), unique = TRUE), 1)
  .dat <- as_tibble(as.data.frame(.data))
  .dat[[nam]] <- seq(nrow(.data))
  dat <- distinct_(.dat, ...)
  
  out <- .data[dat[[nam]], ] 
  #out[[nam]] <- NULL
  out
}

#' @rdname dplyr-Spatial
#' @param y tbl to join
#' @importFrom dplyr left_join inner_join
#' @inheritParams dplyr::left_join
#' @export
left_join.Spatial <- function (x, y, by = NULL, copy = FALSE, ...) {
  x@data <- left_join(as_tibble(as.data.frame(x)), y, by = by, copy = copy, ...)
  x
}

#' @rdname dplyr-Spatial
#' @export
 inner_join.Spatial <- function (x, y, by = NULL, copy = FALSE, ...) {
   randomkey <- paste(sample(c(letters, 1:100)), collapse = "")
   ## kludge to record which rows are kept
   x[[randomkey]] <- seq(nrow(x))
   .data <- inner_join(as_tibble(as.data.frame(x)), y, by = by, copy = copy, ...)
   x <- x[.data[[randomkey]], ]
   .data[[randomkey]] <- NULL
   x@data <- .data
   x
 }





